### PR TITLE
fix(coding-agent): persist kernel stderr to disk and bound the in-memory tail

### PR DESCRIPTION
**Pull request by snimu**
When a Python kernel failed to start, the failure was often undiagnosable: its stderr lived only in an in-memory string that died with the process, and the startup error showed just the last kilobyte of it. Nothing stayed on disk to inspect after the fact — and on the other end, a long-lived chatty kernel grew that same string without bound for its whole lifetime, even though every reader only ever showed the tail.

The log is startup-shaped because the kernel's stderr is startup-shaped: the runtime dup2's fd 2 into its protocol pipe before it reports ready (repl.py `_setup_fds`), so the process-level stderr only ever carries pre-ready bytes — interpreter, venv, and import failures; everything after ready already flows through the protocol into the bounded per-execution stderr. The kernel now always spawns with stderr piped, and the manager's data handler is the single write path into `kernel-stderr.log` in the session's artifact directory: it appends raw chunks under a 5 MiB per-spawn write budget, and once the budget is spent it writes one `[stderr log budget exhausted]` marker, then keeps draining the pipe but discards — so a kernel that spews until the 30s ready timeout (a corrupted venv looping import warnings, multiplied by repair retries) cannot land more than the budget on disk per spawn, and a full pipe can never wedge a pre-ready kernel. The file rotates to `.old` when a spawn finds it over the budget, hard-capping per-session disk at roughly 2× the budget.

Startup failure messages read their 1 KiB tail from the in-memory string — bounded to an 8 KiB tail, decoded across chunk boundaries, and fed by the same pipe as the file, so it sees every logged byte plus host-side `[kernel]` diagnostics. On kernel exit the pipe is drained to quiescence before the stream is torn down, so the kernel's last words reach both the tail and the log even when an orphaned grandchild holds the pipe's write end open past the parent's death. Sessions without an artifact directory (or whose log open fails) keep the memory-tail behavior; the log persists across kernel restarts within the rotation cap.

Validation: three pins — a kernel that writes CR progress output and a split multi-byte character with no trailing newline before dying produces a log byte-identical to what it wrote, with that tail in the startup error; a kernel that spews 6 MiB before dying leaves a log capped at the budget plus the exhaustion marker while the startup error still shows its final words and the script runs to completion (the pipe stayed drained); and a log pre-filled past 5 MiB rotates to `.old` at the next spawn, with the new incarnation's stderr starting a fresh file. Kernel/ipython/repl suites, tsgo, and root `npm run check` green.

Linear: RES-1246 https://linear.app/primeintellect/issue/RES-1246/persist-python-kernel-stderr-to-disk-and-bound-the-in-memory-buffer

## LOC

Total src: **+123/−6 (net +117)**; tests: +244/−3 (net +241).

> [!NOTE]
> ### Persist kernel stderr to disk with per-spawn budget and bound the in-memory tail
> - Adds an optional `stderrLogPath` to `KernelManagerOptions`; when set, `ReplKernelManager.doStart` opens an append log and rotates the prior file when it exceeds the size threshold.
> - Bounds the retained in-memory stderr tail to a configured character limit via `appendKernelStderrText` and `appendKernelDiagnostic`, replacing unbounded accumulation for the kernel lifetime.
> - Writes raw stderr bytes to the log up to a per-spawn budget, emits a one-time exhaustion marker, and continues draining the stream through process exit so final output is not lost.
> - `IpythonKernelProvisioner.startKernel` configures the log path under the session snapshot/artifact directory when it exists.
> - `shutdown` and the pre-ready exit path in `waitForReady` now wait for stderr stream closure before finalizing, instead of discarding buffered data.
> - Risk: `ReplKernelManager.shutdown` no longer destroys the child stderr stream immediately; any caller relying on synchronous stderr teardown during shutdown may see delayed closure until the exit-drain completes.
>
> 
> #### Changes since #1947 opened
>
> - Wrapped stderr log file rotation in `ReplKernelManager.openStderrLogFd` with try/catch error handling that logs a diagnostic message and continues appending to the existing log file when rotation fails, instead of throwing an error [86de757]
> - Added 'end' and 'close' event listeners to `child.stderr` in `ReplKernelManager.start` to flush buffered partial UTF-8 character data from the `TextDecoder` and close the stderr log file descriptor [86de757]
> - Added a new test validating that stderr logging continues when rotation fails, and updated the existing stderr bytes test to verify partial UTF-8 character handling [86de757]
> - Extended the mocked `child.stderr` type in the `configuredManager` test utility to include a `once` method implementation [86de757]
> - Reworked stderr stream lifecycle management in `ReplKernelManager` kernel process exit handler [78a1822]
> - Added tests for `ReplKernelManager` stderr handling edge cases [78a1822]
> - Added `writeFullySync` utility function and replaced direct `fs.writeSync` calls in `ReplKernelManager` kernel stderr data handler with `writeFullySync` to handle partial writes, including writing the budget-exhausted marker as a Buffer [5698258]
> - Extended `node:fs` mock in `repl-kernel-startup.test.ts` to simulate short writes and updated test assertions to verify byte fidelity under partial write conditions [5698258]
> - Modified `ReplKernelManager.openStderrLog` to return both file descriptor and remaining write budget based on existing log file size, and implemented rotation logic that attempts to move oversized logs to `.old` before resetting budget [2a056a9]
> - Updated `ReplKernelManager.start` to track stderr write budget from log file initialization and enforce capacity limits by writing a budget-exhausted marker and discarding further pre-ready stderr bytes when exceeded [2a056a9]
> - Fixed `ReplKernelManager.shutdown` to conditionally destroy child stderr stream only when the child process has not exited, preserving buffered output for already-exited processes [2a056a9]
> - Added test in `repl-kernel-shutdown.test.ts` verifying that `kill()` destroys stderr only when the child process is alive and not when it has an exitCode [2a056a9]
> - Updated tests in `repl-kernel-startup.test.ts` to verify that rotation-failure scenarios respect capacity limits and that write budget equals remaining file capacity at MAX size [2a056a9]
> 
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized e9a7d24.</sup>
> 
>

---

> [!NOTE]
> **Medium Risk**
> Changes kernel subprocess stderr lifecycle and synchronous file I/O on every spawn; behavior is heavily tested but affects startup-failure diagnostics and teardown timing.
> 
> **Overview**
> Pre-ready Python kernel stderr is now persisted and memory-safe instead of living only in an unbounded in-memory string.
> 
> **`ReplKernelManager`** accepts optional `stderrLogPath` and, when set, appends raw pre-ready stderr to `kernel-stderr.log` with rotation to `.old` when the file exceeds 5 MiB, a per-spawn write budget based on remaining file capacity, and continued pipe draining after the budget is exhausted (with a one-time marker) so spew cannot wedge startup. The in-memory diagnostics tail is capped at 8 KiB with UTF-8 decoding across chunk boundaries.
> 
> Startup failures wait for stderr stream closure before building the error tail; exit/teardown avoids destroying stderr on already-exited children so last bytes are not dropped, while still destroying the pipe for live children that ignore kill.
> 
> **`IpythonKernelProvisioner`** sets the log path under the session artifact directory when a snapshot dir exists. Tests cover byte fidelity, rotation, budget caps, rotation failure, grandchild stderr noise, and shutdown stderr destroy rules.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2a056a980e207ad3d9c4e435eccd9d99d17300b7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by xeophon — DISMISSED** [review]

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 1]
```diff
@@ -329,7 +348,7 @@ export class ReplKernelManager {
 		});
 
 		child.stderr?.on("data", (buf: Buffer) => {
-			this.kernelStderr += buf.toString();
+			this.appendKernelStderr(buf.toString());
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts]: Fixed in cf6c42aa6: the in-memory tail now decodes through one persistent StringDecoder per child (mirroring the stdout path), and the disk file sidesteps decoding entirely — raw Buffer chunks are appended verbatim, so a multi-byte character split across chunks survives both surfaces. Pinned with a chunk-split UTF-8 sequence: the disk file byte-equals the exact stream and the tail contains the intact character.

File: packages/coding-agent/src/core/tools/ipython.ts [hunk 2]
```diff
@@ -471,6 +472,7 @@ export class IpythonKernelProvisioner {
 				snapshot: snapshotDir
 					? { path: snapshotPathIn(snapshotDir), manifestPath: manifestPathIn(snapshotDir) }
 					: undefined,
+				stderrLogPath: snapshotDir ? join(snapshotDir, "kernel-stderr.log") : undefined,
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/tools/ipython.ts:478]: Fixed in cf6c42aa6 at the premise: the line-oriented appendRotatingLog is no longer involved. A small private raw sink appends Buffer chunks verbatim — no newline injection, CR-only progress output preserved. Pinned: the disk file byte-equals the exact bytes written across mid-line and mid-character chunk splits.

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 3]
```diff
@@ -201,7 +213,14 @@ export class ReplKernelManager {
...
 
 	private appendKernelDiagnostic(message: string): void {
-		this.kernelStderr += `[kernel] ${message.endsWith("\n") ? message : `${message}\n`}`;
+		this.appendKernelStderr(`[kernel] ${message.endsWith("\n") ? message : `${message}\n`}`);
+	}
+
+	private appendKernelStderr(chunk: string): void {
+		this.kernelStderr = (this.kernelStderr + chunk).slice(-MAX_KERNEL_STDERR_CHARS);
+		if (this.options.stderrLogPath) {
+			appendRotatingLog(this.options.stderrLogPath, chunk.endsWith("\n") ? chunk.slice(0, -1) : chunk);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts]: Fixed in cf6c42aa6: no synchronous fs remains on the stderr data path. The handler stays synchronous and only chains a promise onto a per-sink queue over fs/promises (ordering guaranteed across rotation by the same chain); a failing write is swallowed and the cached size re-stated on the next append. Rotation keeps the existing 5 MiB single-generation policy.

File: packages/coding-agent/src/core/kernel/repl-manager.ts
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts:1]: Fixed in 2b0f40f31: the flush is hoisted into a try/finally wrapping the whole shutdown body, so every resolution kind — the early-return branch after an unexpected exit, thrown teardown errors, and the full path — awaits the sink exactly once (concurrent shutdown() callers share the operation promise, which now resolves after the flush). An exit-time flush was considered and rejected: the exit handler is fire-and-forget and can run before the final stderr data events are delivered, so it guarantees nothing; flush only has meaning as shutdown()'s await barrier. Mutant-checked: with the flush no-op'd, the rotation pin fails deterministically; both pins now assert file contents synchronously after shutdown().

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 4]
```diff
@@ -328,8 +388,10 @@ export class ReplKernelManager {
 			}
 		});
 
+		const stderrDecoder = new StringDecoder("utf8");
 		child.stderr?.on("data", (buf: Buffer) => {
-			this.kernelStderr += buf.toString();
+			this.appendKernelStderrText(stderrDecoder.write(buf));
+			this.stderrSink?.append(buf);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts]: Stale as of the rebuild (b951f55a0): there is no stderrSink and no host-side relay anymore — the child's stderr slot IS the log file descriptor, so a crashed kernel's bytes are syscall-complete in the file when 'exit' fires; destroying child.stderr cannot discard anything because process-level stderr no longer flows through a pipe (the pipe path remains only for sessions without an artifact directory, where behavior is unchanged from before this PR). The byte-fidelity pin asserts file completeness immediately after shutdown() resolves with no wait/flush choreography.

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 5]
```diff
@@ -201,7 +215,45 @@ export class ReplKernelManager {
...
+		const path = this.options.stderrLogPath;
+		if (!path) return undefined;
+		try {
+			mkdirSync(dirname(path), { recursive: true });
+			if (existsSync(path) && statSync(path).size > MAX_KERNEL_STDERR_LOG_BYTES) {
+				// Drop any prior .old first: rename fails on Windows if it exists.
+				rmSync(`${path}.old`, { force: true });
+				renameSync(path, `${path}.old`);
+			}
+			return openSync(path, "a");
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts]: Declining the in-window cap, with the exposure quantified and one adjacent fix shipped (ca7f26ca5). The growth window is not the kernel's lifetime: repl.py's _setup_fds dup2's the protocol pump over fd 2 and closes the log fd BEFORE ready is sent, so the file is frozen ~100-300ms after spawn — a healthy startup writes ~0 bytes and a broken venv writes one traceback and exits. The pathological case (an interpreter spewing at disk speed without exiting or becoming ready) is time-bounded by the 30s READY_TIMEOUT -> reject -> shutdown -> SIGTERM chain, and per-spawn rotation keeps steady-state disk at <= 2 bursts (pre-PR the same adversary filled host RAM with no time bound; the pipe shape suggested here is the sink design deleted after three failed rounds for exactly the defects it reintroduces). There is also no existing pre-ready periodic timer to piggyback a size check on — a mid-window cap would mean a new poller for a never-observed adversary. What the quantification DID expose and fix: stderrTail() read the WHOLE file to produce a 1KB tail, so the failure-reporting path itself would buffer a multi-GB burst in host memory; it now does a positional read of the last 1KB, making host behavior indifferent to log size. 45/45 kernel suite + test:kernel green.

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 6]
```diff
@@ -328,8 +368,47 @@ export class ReplKernelManager {
...
+				} else {
+					writeSync(stderrLogFd, KERNEL_STDERR_LOG_BUDGET_MARKER);
+					stderrLogWritable = false;
+				}
+			} catch (error) {
+				stderrLogWritable = false;
+				this.appendKernelDiagnostic(`kernel stderr log write failed: ${errorMessage(error)}`);
+			}
+		});
+		if (stderrLogFd !== undefined) {
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts]: Confirmed and fixed in 4445e904c. The stderr StringDecoder is now flushed (`decoder.end()`) into the in-memory tail on both stream `end` and `close` — each can be the only event to precede the tail build: on natural EOF the stream's queued `close` emission can land after `exit` (observed on Linux), while a stream destroyed by the exit-drain never emits `end`; the second flush is a no-op. The truncated final character now surfaces as U+FFFD in the startup error tail; the on-disk log keeps the raw bytes either way. Pinned by extending the exact-bytes startup test with a trailing lone `\303` before exit (verified failing without the flush, on Linux and macOS).

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 7]
```diff
@@ -201,7 +216,32 @@ export class ReplKernelManager {
...
+			mkdirSync(dirname(path), { recursive: true });
+			if (existsSync(path) && statSync(path).size > MAX_KERNEL_STDERR_LOG_BYTES) {
+				// Drop any prior .old first: rename fails on Windows if it exists.
+				rmSync(`${path}.old`, { force: true });
+				renameSync(path, `${path}.old`);
+			}
+			return openSync(path, "a");
+		} catch (error) {
+			this.appendKernelDiagnostic(`cannot open kernel stderr log: ${errorMessage(error)}`);
+			return undefined;
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts:259]: Confirmed and fixed in 86de757e6. Rotation now has its own `try`: a failed rm/rename records a `[kernel] cannot rotate kernel stderr log: ...` diagnostic and the manager keeps appending to the existing oversized file, so a locked `.old` never costs the log. Only a failed `openSync`/`mkdirSync` still falls back to the memory-only tail. Pinned with a startup test that plants a directory at the `.old` path (rm/rename both throw) and asserts the new spawn's stderr still lands in the log.

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 8]
```diff
@@ -328,8 +373,54 @@ export class ReplKernelManager {
...
+			if (stderrLogFd !== undefined) closeSync(stderrLogFd);
+		});
+		child.once("exit", () => {
+			// Drain already-buffered stderr to quiescence, then destroy the stream:
+			// its EOF may never come (an orphaned grandchild can inherit the write
+			// end), and destroying at 'exit' would drop the kernel's last words.
+			const drain = () => {
+				const seen = stderrChunks;
+				globalThis.setImmediate(() => {
+					if (stderrChunks !== seen) drain();
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts]: Fixed in 78a1822b3 by re-deriving the teardown shape rather than patching the loop: after the kernel exits, the tail only needs bytes the kernel wrote before dying, and one event-loop turn delivers those from the bounded pipe buffer (libuv drains a readable pipe to EAGAIN within a poll iteration; `setImmediate` runs in the check phase after it). Output arriving later is a surviving grandchild's post-mortem noise, not the kernel's last words, so re-checking for quiescence bought nothing and admitted unbounded deferral in principle. The loop and its chunk counter are deleted; teardown is now a single `setImmediate` then `destroy()`, unconditionally bounded. Empirical note: the literal forever-loop did not reproduce with a busy inherited writer (`while :; do echo ...; done >&2 &`) on Linux or macOS — the writer blocks on the full pipe while node drains it, so quiet rounds occur — but the bound was timing-dependent, which the new shape removes. Pinned: a kernel that leaves a never-quiet, never-EOF stderr writer behind still gets its startup failure reported promptly (destroying the pipe ends the writer with SIGPIPE).

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 9]
```diff
@@ -328,8 +373,54 @@ export class ReplKernelManager {
...
+		});
+		// A kernel that dies mid-character leaves bytes buffered in the decoder;
+		// flush them so the tail keeps the truncated final character. Both events,
+		// because each can be the only one to precede the tail build: 'end' beats
+		// 'exit' on natural EOF (whose 'close' emission can land after it), while
+		// a drain-destroyed stream skips 'end'. The second end() returns "".
+		child.stderr?.once("end", () => this.appendKernelStderrText(stderrDecoder.end()));
+		child.stderr?.once("close", () => {
+			this.appendKernelStderrText(stderrDecoder.end());
+			if (stderrLogFd !== undefined) closeSync(stderrLogFd);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts]: Confirmed and fixed in 78a1822b3. `closeSync(stderrLogFd)` in the stream's `close` listener is now wrapped: a throw there was an uncaught exception in an event handler and would have terminated the host, inconsistent with open/rotate/write failures which all degrade with a diagnostic. Close failures now append a `[kernel] kernel stderr log close failed: ...` diagnostic instead. Pinned with a scoped `node:fs` mock that marks the log fd at open and makes exactly its close throw (there is no way to force a real `closeSync` failure): the startup error still carries the kernel's stderr plus the close diagnostic, and no unhandled error escapes; without the guard the pin dies of the unhandled exception and rides the full ready timeout.

File: packages/coding-agent/src/core/kernel/repl-manager.ts [hunk 10]
```diff
@@ -328,8 +373,51 @@ export class ReplKernelManager {
...
+		const stderrLogFd = this.openStderrLogFd();
+		const stderrDecoder = new StringDecoder("utf8");
+		let stderrLogBudget = MAX_KERNEL_STDERR_LOG_BYTES;
+		let stderrLogWritable = stderrLogFd !== undefined;
 		child.stderr?.on("data", (buf: Buffer) => {
-			this.kernelStderr += buf.toString();
+			this.appendKernelStderrText(stderrDecoder.write(buf));
+			if (!stderrLogWritable || stderrLogFd === undefined) return;
+			try {
+				if (buf.length <= stderrLogBudget) {
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/kernel/repl-manager.ts:399]: Confirmed and fixed in 56982582b. Chunk and marker writes now go through a write-all loop (`writeFullySync`: repeat `writeSync` from the advancing offset until the buffer is on disk), so a short write can no longer drop the suffix; on success the budget decrement equals the bytes actually written, and a mid-loop throw still lands in the existing catch that disables logging with a diagnostic. Pinned inside the byte-exact startup test: the test file's `node:fs` mock factory halves every write to the log fd (marked at open by path) while that test is armed, so only a write-all loop reproduces the kernel's stderr byte-for-byte — verified failing on the previous single-write code.

**Review by sethkarten — APPROVED** [review]
